src/System.Diagnostics.StackTrace.Symbols/src/System/Diagnostics/StackTrace/Symbols.cs

```diff
@@ -0,0 +1,222 @@
+                                        documents[i] = sequence.Document;
+                                        i++;
+                                    }
+
+                                    // Search for the correct IL offset
+                                    int j;
+                                    for (j = 0; j < sequencePointCount; j++) {
+
+                                        // look for the entry matching the one we're looking for
+                                        if (offsets[j] >= ilOffset) {
+
+                                            // if this offset is > what we're looking for, ajdust the index
+                                            if (offsets[j] > ilOffset && j > 0) {
+                                                j--;
+                                            }
+                                            break;
+                                        }
+                                    }
+
+                                    // If we didn't find a match, default to the last sequence point
+                                    if (j == sequencePointCount) {
+                                        j--;
+                                    }
+
+                                    while (lines[j] == SequencePoint.HiddenLine && j > 0) {
+                                        j--;
+                                    }
+
+                                    if (lines[j] != SequencePoint.HiddenLine) {
+                                        sourceLine = lines[j];
+                                        sourceColumn = columns[j];
+                                    }
+                                    sourceFile = reader.GetString(reader.GetDocument(documents[j]).Name);
+                                }
+                            }
+                            catch {
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the portable PDB reader for the assembly path
+        /// </summary>
+        /// <param name="assemblyPath">file path of the assembly or null</param>
+        /// <param name="inMemorySymbols">in memory PDB address or zero</param>
+        /// <param name="inMemorySymbolsSize">in memory PDB size</param>
+        /// <param name="reader">returns the reader</param>
+        /// <returns>provider</returns>
+        static IDisposable GetReader(string assemblyPath, IntPtr inMemorySymbols, int inMemorySymbolsSize, out MetadataReader reader)
+        {
+            reader = null;
+            MetadataReaderProvider provider = null;
+
+            if (assemblyPath != null) {
+                try {
+                    uint stamp;
+                    int age;
+                    Guid guid;
+
+                    string pdbName = GetPdbPathFromPeStream(assemblyPath, out age, out guid, out stamp);
```

Comment:
This is guaranteed to be a perf problem - opening and closing two files per a line in the stacktrace.

Can we at least design the API so that it is caching friendly? E.g. instead of static method, have SymbolProvide object - the runtime will create instance of it before building the stacktrace, ask it to resolve the frames, and then dispose it.